pkg/bspatch/bspatch.go

```diff
@@ -100,141 +125,132 @@ func patchb(oldfile, patch []byte) ([]byte, error) {
-			e0 := ""
-			if err != nil {
-				e0 = err.Error()
-			}
-			return nil, fmt.Errorf("corrupt patch or bzstream ended (2): %s", e0)
+		if newfwc.Count()+ctrip[x] > newsize {
+			return newCorruptPatchError("newfile pos + data block exceeds expected newfile size")
 		}
-		// Add pold data to diff string
-		for i = 0; i < ctrl[0]; i++ {
-			if oldpos+i >= 0 && oldpos+i < oldsize {
-				pnew[newpos+i] += oldfile[oldpos+i]
-			}
+
+		// Read x bytes from diff + old into new file
+		lenread, err = io.CopyBuffer(newfwc, io.LimitReader(xbyteadd, ctrip[x]), cpBuf)
+		if lenread < ctrip[x] || (err != nil && err != io.EOF) {
+			return newCorruptPatchBzEndError(lenread, ctrip[x], "x data block", err)
 		}
 
-		// Adjust pointers
-		newpos += ctrl[0]
-		oldpos += ctrl[0]
+		if newfwc.Count()+ctrip[y] > newsize {
+			return newCorruptPatchError("newfile pos + extra block exceeds expected newfile size")
+		}
 
-		// Sanity-check
-		if newpos+ctrl[1] > newsize {
-			return nil, fmt.Errorf("corrupt patch newpos+ctrl[1] newsize")
+		// Read y bytes from the extra block into the new file
+		lenread, err = io.CopyBuffer(newfwc, io.LimitReader(xtra, ctrip[y]), cpBuf)
+		if lenread < ctrip[y] || (err != nil && err != io.EOF) {
+			return newCorruptPatchBzEndError(lenread, ctrip[y], "y extra block", err)
 		}
 
-		// Read extra string
-		// epfbz2.Read was not reading all the requested bytes, probably an internal buffer limitation ?
-		// it was encapsulated by zreadall to work around the issue
-		lenread, err = zreadall(epfbz2, pnew[newpos:newpos+ctrl[1]], ctrl[1])
-		if lenread < ctrl[1] || (err != nil && err != io.EOF) {
-			e0 := ""
-			if err != nil {
-				e0 = err.Error()
-			}
-			return nil, fmt.Errorf("corrupt patch or bzstream ended (3): %s", e0)
+		// Adjust oldfile offset by z
+		_, err = oldf.Seek(ctrip[z], os.SEEK_CUR)
+		if err != nil {
+			return err
 		}
-		// Adjust pointers
-		newpos += ctrl[1]
-		oldpos += ctrl[2]
 	}
 
 	// Clean up the bzip2 reads
-	if err = cpfbz2.Close(); err != nil {
-		return nil, err
+	if err = ctrl.Close(); err != nil {
+		return err
 	}
-	if err = dpfbz2.Close(); err != nil {
-		return nil, err
+	if err = data.Close(); err != nil {
+		return err
 	}
-	if err = epfbz2.Close(); err != nil {
-		return nil, err
+	if err = xtra.Close(); err != nil {
+		return err
 	}
-	cpfbz2 = nil
-	dpfbz2 = nil
-	epfbz2 = nil
-	cpf = nil
-	dpf = nil
-	epf = nil
-
-	return pnew, nil
+	ctrlbz = nil
+	databz = nil
+	xtrabz = nil
+	ctrl = nil
+	data = nil
+	xtra = nil
+
+	return nil
+}
+
+func patchb(oldfile, patch []byte) ([]byte, error) {
+	newfby := new(bytes.Buffer)
+	// Use bufio here to emulate File()'s use of bufio for testing
+	newfbuf := bufio.NewWriterSize(newfby, writeBufferSize)
+	oldfby := bytes.NewReader(oldfile)
+	err := patchStream(oldfby, newfbuf, patch)
+	newfbuf.Flush()
+	return newfby.Bytes(), err
 }
 
 // offtin reads an int64 (little endian)
-func offtin(buf []byte) int {
+func offtin(buf []byte) int64 {
```

Comment:
Get rid of this function, and just use `int64(binary.LittleEndian.Uint64(buf))`.

I'm not convinced this function is bug-free either -- it seems to be parsing a ones' complement signed int (as opposed to two's complement). I'm not sure what the spec says. That being said, I don't think this number would ever be negative in practice.

Reply:
So it's actually possible for it to be negative, and the `bsdiff` / `bspatch` implementation in this repo both use ones' complement. It's probably safer to just leave as-is.